### DISCUSSION TITLE
JBang vs Pip

### DISCUSSION THREAD

**Discussion by aghasemi** [jbangdev/jbang#1250] [General]
Hi,

Sorry if the question looks trivial, but I thought someone else may have this slight confusion:

So, in Python, if I have a command-line script, I can package it with a `setup.py` and then install it with `pip3 install mypackage`, which takes care of depdendencies, adding to PATH, etc... My questions is, what is the equivalent of this (i.e. installing a CLI script globally in the system) for JBang?

JBang, as far as I know, can run scripts directly from web e.g. a GitHub link. In that sense, is the script cached somewhere, or it is downloaded before every execution?

In the case of app store, is the script cached, and can I access it with a shorter identifier than what I use at the time of installing? 

Thanks a lot

**Comment by quintesse** [comment]
You can install scripts with JBang in a mostly similar way: `jbang app install <script>`. There's no setup, if you can run it with `jbang <script>`, you can also install it with `jbang app install <script>`. JBang will try to guess a name, which you can override with `--name`, eg: `jbang app install --name myscript somescript.java`. After that you can simply run `myscript`.

And scripts are indeed cached, as are the compilation results and things like that. (You can force JBang to download/compile again using `--fresh`, eg: `jbang --fresh <script>`).

And for being able to reference scripts using shorter identifiers, that's indeed possible using `jbang alias add`, eg: `jbang alias add --name myname http://some.server/with/a/very/long/path/to/a/file.java`. You can then simply run using `jbang myname`.

HTH

**Reply by aghasemi** [reply]
Many thanks. I found this section in docs: https://www.jbang.dev/documentation/guide/latest/install.html.
Is this different from the JBang AppStore (https://www.jbang.dev/appstore/) where one has to set up a `catalog.json` file or `jbang-catalog` repository?

**Reply by quintesse** [reply]
That's more for sharing your scripts with others. 

When you use `jbang alias add`, those aliases gets added to a local `jbang-catalog.json`. So when you share a catalog like that (via the app store, jbang-catalog repository or any other means) it's like sharing a list of aliases. Which makes it very easy for others to use the scripts you've made.

Does that make sense?

**Reply by aghasemi** [reply]
It does, I think. 

So, if I want to use a Jbang script as a command in my own system as well as share it for others, the minimal thing to do is to put it in a GitHub repository and install it using `jbang app install` with the name of the command being specified in the `--name` parameter, right?

The `jbang-catalog` comes in handy e.g. when I have a collection of possibly related commands then. No?

**Reply by quintesse** [reply]
Indeed. There are some examples of catalogs that have useful commands related to their project. But there are also some catalogs (like mine), that are basically a hodge-podge of different experiments :-)

**Reply by aghasemi** [reply]
Many thanks.

**Reply by maxandersen** [reply]
Just to be precise jbang install can "install" anything jbang can run. Ie. A script, jar and dependency stored locally or on any set of the remote adresses it supports. For example you can share script to anyone as long as it as available via http/https 

The catalog purpose is twofold. Works like an index that tools or humans can use and to get shorter hand names. Ie. Env@jbangdev instead of the absolute url the env script actually is stored at. 

So jbang catalogs are not required but it just makes things more discoverable and easier to refer to. 

Hope that helps.
